npm/vite-dev-server/src/plugins/cypress.ts

```diff
@@ -52,21 +53,45 @@ export const Cypress = (
     configResolved (config) {
       base = config.base
     },
-    async transformIndexHtml () {
+    async transformIndexHtml (html) {
+      // it's possibe other plugins have modified the HTML
+      // before we get to. For example vitejs/plugin-react will
+      // add a preamble. We do our best to look at the HTML we
+      // receive and inject it.
+      // For now we just grab any `<script>` tags and inject them to <head>.
+      // We will add more handling as we learn the use cases.
+      const root = parse(html)
```

Comment:
So, what this is doing is basically grabbing any `<script>` tags from the `html` that might have been inserted by other plugins and sticking them inside our HTML. An example is the vitejs/react-plugin, which injects this code: https://github.com/vitejs/vite/blob/main/packages/plugin-react/src/fast-refresh.ts#L32-L36. A user already reported this: https://github.com/vitejs/vite/issues/9362.

Reply:
Had a similar instinct, but I ran into a few roadblocks with this. With regards to using the `{ html, tags }` API for the HTML, by default we inject something like this:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <title>Components App</title>
  </head>
  <body>
    <div data-cy-root></div>
  </body>
</html>
```

Let's call this `baseHtml`. Users can customize this. We could do:

```js
return {
  html: baseHtml,
  tags: /* tags to inject */
}
```

Injecting the `script` content is pretty awkward, though, since we need to parse the `attrs` to use the `tags` API, eg:

```js
const tags = scriptTagsToInject.map<HtmlTagDescriptor>(script => ({
  tag: 'script',
  injectTo: 'head',
  attrs: {
    // is it src="..." or module="..."
  }
}))

return {
  html: newHtml,
  tags
}
```

So we need to parse the tag in even more detail. For many plugins, this API works great, like `plugin-react`, since they know they want to inject `type="module"`: [see here](https://github.com/vitejs/vite/blob/main/packages/plugin-react/src/index.ts#L390-L392). For us to use this API, we need to completely pass everything, which can include `<script>`, `<div>`, etc - this seems challenging and prone to error.

I could be missing something, it would be much nicer to use the Vite API than string substitution, definitely open to improving this if you see some alternatives. Right now there are definitely edge case, for example, if another plugin injects a `<div>` or something in `<head>` we won't grab it.

Reply:
Yuck. Makes sense to just stick with the string sub for now, adding that level of additional parsing feels very brittle and error-prone